interfaces/input_parameters/ring_section.py

```diff
@@ -1,77 +1,260 @@
+                        'as non-linear momentum compaction factor. ' +
+                        'The correct syntax is alpha_n.')
+
+        # Setting all the valid non-linear alpha and replacing
+        # undeclared orders with zeros
+        self.alpha_orders_defined = [0]
+        for order in range(1, self.alpha_order+1):
+            alpha = alpha_n.pop(order, None)
+
+            if alpha is None:
+                alpha = 0
+                # This condition can be replaced by 'continue' to avoid
+                # populating the object with 0 programs
+            else:
+                self.alpha_orders_defined.append(order)
+
+            if not isinstance(alpha, dTypes.momentum_compaction):
+                alpha = dTypes.momentum_compaction(alpha, order=order)
+            else:
+                if alpha.order != order:
+                    raise excpt.InputError(
+                        "The order of the datatype passed as keyword " +
+                        "argument alpha_%s do not match" % (order))
+
+            self._check_and_set_momentum_compaction(alpha, order)
+
+    def _check_and_set_momentum_compaction(self, alpha, order):
+        '''
+        Internal function to check that the input momentum compaction is
+        coherent with the synchronous data. If the synchronous data is turn
+        based, the momentum compaction should have the same length. If the
+        synchronous is time based while the momentum compaction is turn based,
+        raises a warning.
+        '''
+
+        setattr(self, 'alpha_'+str(order), alpha)
+
+        if self.synchronous_data.ndim == 2:
 
-        if not hasattr(alpha, '__iter__'):
-            alpha = (alpha, )
-        
-        if isinstance(alpha, dict):
             try:
-                if not all([k%1 == 0 for k in alpha.keys()]):
-                    raise TypeError
-            except TypeError:
-                raise excpt.InputError("If alpha is dict all keys must be "
-                                       + "numeric and integer")
-
-            maxAlpha = np.max(tuple(alpha.keys())).astype(int)
-            alpha = [alpha.pop(i, 0) for i in range(maxAlpha+1)]
-        
-        if isinstance(alpha, dTypes._function) or \
-            isinstance(alpha, bf.machine_program):
-            alpha = (alpha,)
-
-        for i, a in enumerate(alpha):
-            if not isinstance(a, dTypes.momentum_compaction):
-                a = dTypes.momentum_compaction(a, order = i)
-            setattr(self, 'alpha_'+str(i), a)
-        self.alpha_order = i
-        
-        for i in range(3 - self.alpha_order):
-            if not hasattr(self, f'alpha_{i}'):
-                setattr(self, 'alpha_'+str(i), dTypes.momentum_compaction(0,
-                                                                  order = i))
+                setattr(self, 'alpha_'+str(order), alpha.reshape(
+                    1, use_turns=np.arange(self.synchronous_data.shape[1])))
```

Comment:
This will only work if `interp_time == t_rev`.  It's sufficient to call it with `use_turns = self.use_turns, use_time = self.cycle_time` and then the check for which is needed is handled internally

Reply:
Function adapted to better use datatypes properties